### PR TITLE
linker: add iterable section macros

### PR DESCRIPTION
**Pull request by tyrantlizard77**
For iterable areas defined with Z_STRUCT_SECTION_ITERABLE(),
the corresponding output section in the linker script is just
boilerplate. Add macros to make these definitions simpler.

Unfortunately, we have a fair number of iterable sections not
defined with Z_STRUCT_SECTION_ITERABLE(), this patch does not
address this.

The output sections are all named `<struct name>_area`, update
sanitylib.py with this.

sys_sem with no userspace, and k_lifo/k_fifo are special cases
where different data types that are all equivalent need to be
put in the same iterable area. Add
Z_STRUCT_SECTION_ITERABLE_ALTERNATE() for this special case.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Review by npitre — APPROVED** [review]

**Comment by andrewboie** [comment]
Checkpatch issues are false positives, we are using the preprocessor here for something that is not C code.

File: include/linker/common-rom.ld [hunk 1]
```diff
@@ -118,36 +108,16 @@
...
 
-	SECTION_DATA_PROLOGUE(_bt_channels_area,,SUBALIGN(4))
-	{
-		_bt_l2cap_fixed_chan_list_start = .;
-		KEEP(*(SORT_BY_NAME("._bt_l2cap_fixed_chan.static.*")))
-		_bt_l2cap_fixed_chan_list_end = .;
-	} GROUP_LINK_IN(ROMABLE_REGION)
+	Z_ITERABLE_SECTION_ROM(bt_l2cap_fixed_chan, 4)
 
 #if defined(CONFIG_BT_BREDR)
```
Comment (andyross) [inline include/linker/common-rom.ld:113]: For all these, it seems like the #ifdef is needless.  There's no alignment being done or anything else that will be visible outside the link (beyond the existence of the start/end symbols).  Seems like we should be able to include these under all conditions.

I guess the argument against would be "what if there was stuff in the "_bt_l2cap_br_fixed_chan" section when CONFIG_BT_BREDR is unset?".  In which my answer would be: isn't that a bug already?

File: include/toolchain/common.h [hunk 2]
```diff
@@ -180,11 +180,22 @@
...
  */
 #define Z_STRUCT_SECTION_ITERABLE(struct_type, name) \
 	Z_DECL_ALIGN(struct struct_type) name \
 	__in_section(_##struct_type, static, name) __used
 
+/* Special variant of Z_STRUCT_SECTION_ITERABLE, for placing alternate
+ * data types within the iterable section of a specific data type. The
+ * data type sizes and semantics must be equivalent!
+ */
+#define Z_STRUCT_SECTION_ITERABLE_ALTERNATE(out_type, struct_type, name) \
```
Comment (andyross) [inline include/toolchain/common.h:195]: Seems like we should be able to fix the naming and init conventions for sem and queue/fifo/lifo such that this isn't needed, but it's a 2-line internal API so maybe not worth worrying about.

**Review by andyross — APPROVED** [review]
Much cleaner.  Some quibbles about stuff to think about.

**Review by carlescufi — COMMENTED** [review]
This PR seems to reorder the handles of the services completely (i.e. the service that has the lowest handle number in master has the highest in this PR).
@Vudentz could you take a look?

**Comment by andrewboie** [comment]
> This PR seems to reorder the handles of the services completely (i.e. the service that has the lowest handle number in master has the highest in this PR).
> @Vudentz could you take a look?

I think I know what this is. I dropped the `SORT` directives in the macros, but it seems this is necessary for this piece. Thanks for helping check this out!

File: include/linker/common-rom.ld (hunk 1 above)
Comment (andrewboie) [reply include/linker/common-rom.ld:113]: Looking at this, yes I agree there are a bunch of conditional ifdefs in these files (not just CONFIG_BT_BREDR, there are many more) that could be removed with no ill effects. Should be in a separate patch though.

**Pull request merged**